src/bit-fields/src/lib.rs

```diff
@@ -0,0 +1,735 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! A macro to generate structures which support bit flags and sub-bytes ranges.
+//!
+//! A [bitflags](https://crates.io/crates/bitflags) like library which also supports value ranges.
+//!
+//! Originally designed to support efficiently interacting with [CPUID](https://en.wikipedia.org/wiki/CPUID).
+//!
+//! See [`ExampleBitField`] for an example.
+//!
+//! ## Features
+//! - `set_theory`: Set theory functions, these are `superset`, `subset`, `disjoint`, `intersection`
+//!   and `union`.
+//! - `bit_index`: Bit indexing, allowing constant indexing like `.bit::<3>()` and
+//!   `.bit_mut::<3>()`.
+//! - `flag_set`: Flag set interoperability, allowing converting to and form a hashset of strings
+//!   referring to bit flags.
+//! - `field_map`: Field map interoperability, allowing converting to and from a hashmap of string
+//!   field names and data type field values.
+//! - `display`: `std::fmt::Display` implementation.
+//! - `cmp_flags`: The `cmp_flags` function which can be used to compare bit flags of 2 bit field
+//!   structures.
+//! - `serde`: Serde serialize and deserialize implementations.
+
+#![warn(clippy::pedantic, clippy::restriction)]
+// The only way I can see the enums changing would be changes to the variants themselves rather than
+// addition of new variants. This means whether they non-exhaustive or not would not matter as the
+// changes would be breaking regardless.
+#![allow(
+    clippy::blanket_clippy_restriction_lints,
+    clippy::implicit_return,
+    clippy::exhaustive_enums,
+    clippy::decimal_literal_representation,
+    clippy::non_ascii_literal,
+    clippy::std_instead_of_core,
+    clippy::missing_docs_in_private_items,
+    clippy::std_instead_of_alloc,
+    clippy::pub_use,
+    clippy::single_char_lifetime_names,
+    clippy::wildcard_imports,
+    clippy::module_name_repetitions,
+    clippy::mod_module_files
+)]
```

Comment:
I don't think these should be here when we merge into main. We should keep our clippy lint consistent across the entire workspace imo, and I don't see us enabling all pedantic and restriction lints globally, ever, as that would be too annoying.

Reply:
Stricter linting is an improvement to security and safety.

I don't think we can require improvements only be implemented globally, we need to allow for improvements like this to be implemented incrementally and thus they will not always be consistent across the workspace.

> and I don't see us enabling all pedantic and restriction lints globally, ever, as that would be too annoying.

On this I would disagree, I don't think we can avoid safety improvements because they are slightly annoying. I also typically don't find much trouble resolving the warnings but this will vary and is considerable personal preference.

Reply:
Whether stricter linting improves security and safety very highly depends on what is being linted. In the case of pedantic and restrictive lints, we get a _lot_ of lints completely unrelated to security and safety (for example, those you explicitly allow above) and a lot of lints with false positives (e.g. the allow directives sprinkled throughout this PR). If we enable these, we will end up polluting the code with a LOT of `allow`.

I agree that changes need to happen incrementally, but I would argue it should happen per lint (after discussion about the added value of the lint), but project wide, as otherwise the linting between the individual crates will inevitably diverge, or worse become contradictory.

Reply:
> we will end up polluting the code with a LOT of allow

I wouldn't consider 14 lines per crate a lot.

Reply:
I am not saying the work you did here to make the code safer is not appreciated, or bad! I actually think there is a lot of value is checking for lints that go beyond what our CI blocks a build on. I simply argue against leaving such extensive traces of it in the code base. And I am not just talking about the 14 `allow` directives here, you are right they are not many. I was talking more about the `allow` directives annotating methods/etc. that we have to introduce due to false-positives/lints misfiring. These already at points are out-of-sync with the actual code they annotate. For example [this one](https://github.com/firecracker-microvm/firecracker/blob/c03d244bea745f4c617a125b67e50aab78348813/src/bit-fields-macros/src/builder.rs#L12) is already covered by the global allow at the top of the module. [This function](https://github.com/firecracker-microvm/firecracker/blob/c03d244bea745f4c617a125b67e50aab78348813/src/bit-fields-macros/src/builder.rs#L91) does not use any integer arithmetic and [here](https://github.com/firecracker-microvm/firecracker/blob/c03d244bea745f4c617a125b67e50aab78348813/src/bit-fields-macros/src/utils.rs#L24) no `as_conversions` are used. 

I think here there is more value in not having the `allow` and instead just having comments explaining why things are safe (if its not obvious). These comments would also be closer to the actual line on which the lint fires, instead of being at the top of a (potentially very long) function. But this is probably an item for the decision meeting.

Sorry I didn't make this clearer earlier :(

Reply:
I've removed the unneeded `allow`s you've pointed out.

> I think here there is more value in not having the allow and instead just having comments explaining why things are safe (if its not obvious).

We need the allow otherwise clippy fails with a warning, to be able to remove the allow is to have to remove `clippy::pedantic` and/or `clippy::restriction`.

I would maintain removing these stricter lints is a degradation in code quality, certainly safety.

Reply:
Also [this one](https://github.com/firecracker-microvm/firecracker/blob/c03d244bea745f4c617a125b67e50aab78348813/src/bit-fields-macros/src/builder.rs#L12) is not covered by the global lint. The global allow for this lint is in a different crate.

Reply:
Whoops, sorry, mixed up the crates there!

Anyway, since I learnt that this PR is not to be merged into main yet, whether we want to enable pedantic/restriction level lints is a discussion we can have at a later time, and not a blocker here.